zkevm-circuits/src/evm_circuit/execution/callop.rs

```diff
@@ -217,13 +239,101 @@ impl<F: Field> ExecutionGadget<F> for CallOpGadget<F> {
+                        CallContextFieldTag::CallDataOffset,
+                        call_gadget.cd_address.offset(),
+                    ),
+                    (
+                        CallContextFieldTag::CallDataLength,
+                        call_gadget.cd_address.length(),
+                    ),
+                    (
+                        CallContextFieldTag::ReturnDataOffset,
+                        call_gadget.rd_address.offset(),
+                    ),
+                    (
+                        CallContextFieldTag::ReturnDataLength,
+                        call_gadget.rd_address.length(),
+                    ),
+                ] {
+                    cb.call_context_lookup(
+                        true.expr(),
+                        Some(callee_call_id.expr()),
+                        field_tag,
+                        value,
+                    );
+                }
+
+                // Save caller's call state
+                for (field_tag, value) in [
+                    (CallContextFieldTag::LastCalleeId, callee_call_id.expr()),
+                    (CallContextFieldTag::LastCalleeReturnDataOffset, 0.expr()),
+                    (
+                        CallContextFieldTag::LastCalleeReturnDataLength,
+                        // return_data_len.expr(),
+                        1.expr(), // TODO(rohit): fix this.
+                    ),
+                ] {
+                    cb.call_context_lookup(true.expr(), None, field_tag, value);
+                }
+
+                // copy table lookup to verify the copying of bytes:
+                // - from caller's memory (`call_data_length` bytes starting at `call_data_offset`)
+                // - to the current call's memory (`call_data_length` bytes starting at `0`).
+                cb.condition(call_gadget.cd_address.has_length(), |cb| {
+                    cb.copy_table_lookup(
+                        cb.curr.state.call_id.expr(),
+                        CopyDataType::Memory.expr(),
+                        callee_call_id.expr(),
+                        CopyDataType::Memory.expr(),
+                        call_gadget.cd_address.offset(),
+                        call_gadget.cd_address.address(),
+                        0.expr(),
+                        call_gadget.cd_address.length(),
+                        0.expr(),
+                        2.expr() * call_gadget.cd_address.length(), // reads + writes
+                    );
+                });
+
+                PrecompileGadget::construct(
+                    cb,
+                    call_gadget.is_success.expr(),
+                    call_gadget.callee_address_expr(),
+                    cb.curr.state.call_id.expr(),
+                    call_gadget.cd_address.offset(),
+                    call_gadget.cd_address.length(),
+                    call_gadget.rd_address.offset(),
+                    call_gadget.rd_address.length(),
+                )
+            },
```

Comment:
We'd need to do step state transition here to callee's state here, and restore that in callee's step, right? Otherwise I don't see any constraint to enforce that.